cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -359,34 +365,69 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
-      } else if (runId != null) {
-        ProgramRunId programRunId = programId.run(runId);
-        // Check if the program is running and is started by the Workflow
-        RunRecordMeta runRecord = store.getRun(programId, runId);
+      }
+      throw new BadRequestException(String.format("Program '%s' is not running.", programId));
+    }
+
+    // Stop the running program based on a combination of runtime info and run record
+    // It's possible that some of them are not yet available from the runtimeService due to timing
+    // differences between the run record was created vs being added to runtimeService
+    // So we retry in a loop for up to 3 seconds max to cater for those cases
+
+    Set<String> pendingStops = Stream.concat(runtimeInfos.keySet().stream().map(RunId::getId),
+                                             activeRunRecords.keySet().stream().map(ProgramRunId::getRun))
+                                      .collect(Collectors.toSet());
+
+    List<ListenableFuture<ProgramController>> futures = new ArrayList<>();
+    Stopwatch stopwatch = new Stopwatch().start();
+
+    while (!pendingStops.isEmpty() && stopwatch.elapsedTime(TimeUnit.SECONDS) < 3L) {
+      Iterator<String> iterator = pendingStops.iterator();
+      while (iterator.hasNext()) {
+        ProgramRunId activeRunId = programId.run(iterator.next());
+        RunRecordMeta runRecord = activeRunRecords.getOrDefault(activeRunId,
+                                                                store.getRun(programId, activeRunId.getRun()));
+        // Check if the program is actually started from workflow and the workflow is running
         if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
           && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
           String workflowRunId = runRecord.getProperties().get("workflowrunid");
           throw new BadRequestException(String.format("Cannot stop the program '%s' started by the Workflow " +
-                                                        "run '%s'. Please stop the Workflow.", programRunId,
+                                                        "run '%s'. Please stop the Workflow.", activeRunId,
                                                       workflowRunId));
         }
-        throw new NotFoundException(programRunId);
+
+        RuntimeInfo runtimeInfo = runtimeService.lookup(programId, RunIds.fromString(activeRunId.getRun()));
+        if (runtimeInfo != null) {
+          futures.add(runtimeInfo.getController().stop());
+          iterator.remove();
+        }
+      }
+
+      // If not able to stop all of them. Try to refresh the list.
+      if (!pendingStops.isEmpty()) {
+        pendingStops = Stream.concat(findRuntimeInfo(programId, runId).keySet().stream().map(RunId::getId),
+                                     getActiveRuns(programId, runId).keySet().stream().map(ProgramRunId::getRun))
```

Comment:
If we refresh the list - in the rare scenario where you stop the program and start the program immediately (possible through other thread within 3 seconds) - would it be possible that the newly launched run will also get stopped?

Reply:
True. However, it depends on how we define the "stop" behavior, since the stop call hasn't been returned yet. Basically whether we define it to be taking a snapshot what needs to be stopped and try to stop them, versus, stopping all active runs (which can be a bit more dynamic).

Reply:
sounds good. so we are defining `stop` to be more dynamic.